### DISCUSSION TITLE
[v3] wire:loading.delay element is showed and not removed for validation errors after 1 successful request

### DISCUSSION THREAD

**Discussion by MatinMN** [livewire/livewire#6338] [7. Bugs (archived)]
### Livewire version

3.0.0-beta.8

### Which PHP version are you using?

PHP 8.2

### Repository URL

https://github.com/MatinMN/livewire-delay-bug

### Steps To Reproduce

Wire:loading.delay elements are not removed, if you perform a successful form submission that takes longer than 200ms and subsequently perform a request that has validation errors. Loading element is shown and is not removed (indefinitely shown)
Please refer to the code examples to reproduce.

1. Create a form submission with wire:loading.delay to show loading state and validation or use the demo repository
2. perform a successful form submission with all inputs entered (the form submission must take longer than 200ms to display loading bar)
3. submit an empty form to trigger validation errors
4. wire:loading.delay will show and not get removed 

There was a similar issue reported on wire:loading which was fixed and included on beta.8 however, it does not seem to fix the above issue. 
The above bug only happens when using wire:loading.delay, when using wire:loading everything works as expected.

https://github.com/livewire/livewire/discussions/6040

video on how to Reproduce with demo code in Additional information code
https://github.com/livewire/livewire/assets/14541783/1c7b84da-19a7-4ec2-bd2b-6c022ee4cb56

video from the demo repository code
https://github.com/livewire/livewire/assets/14541783/ec8ed429-ad08-4107-a09d-1d0c76645917

### Additional information

PHP Component Code

```PHP
    public $email;

    public function submitForm(){
        $this->validate([
            'email' => 'required|email|max:255'
        ]);
        
        sleep(1); // or any slow operation / slow connection also works

        $this->reset();
    }

    public function render()
    {
          return view('livewire.contact-us'); // add your component blade file name please
    }
```

Example Blade file ( no css styling included, please don't use tailwind to ensure the input is visible or use repository example which includes tailwind and css styling)

```BLADE
<div>
    <div wire:loading.delay>
        <span>Loading...</span>
    </div>
    <form wire:submit="submitForm" action="#" class="space-y-8">
        <input wire:model='email' type="email" id="email"
            class="">
        @error('email')
            <span class="">{{ $message }}</span>
        @enderror

        <button type="submit"
            class="">Send
            message</button>
    </form>
</div>
```

### Please confirm (incomplete submissions will not be addressed)

- [X] I have provided easy and step-by-step instructions to reproduce the bug.
- [X] I have provided code samples as text and NOT images.
- [X] I understand my bug report will be removed if I haven't met the criteria above.

**Comment by arnaudome** [comment]
I have a problem that, I think, comes from the exact same bug.

According to the docs, if an element has `wire:loading.delay         <!-- 200ms -->`  and the request is under 200 ms it should not appear.

In my case, the request lasts 90 ms but the element with `wire:loading.delay` appears and never disappears.
The element is a loading spinner.

In fact, as said by MartinMN, if I change `wire:loading.delay` to `wire:loading` works fine.

FYI: The button is allocated in a filament Form inside a Livewire component.

**Comment by SeriousKen** [comment]
I'm having a similar problem, I discovered it working with Filament (https://github.com/filamentphp/filament/discussions/7722#discussion-5516981) but this seems to come from a short response time, if I make the response slow then the problem goes away.

**Comment by SeriousKen** [comment]
Done some more on this and get really odd behaviour. Set up a simple counter component as follows:

``` php
<?php

namespace App\Livewire;

use Livewire\Component;

class Counter extends Component
{
    public $value = 0;

    public function render()
    {
        return view('livewire.counter');
    }

    public function update()
    {
        $this->value++;
    }
}
```

With a view:

``` blade
<div>
    <p>You have clicked the button {{ $value }} times</p>
    <button type="button" wire:click='update'>Click</button>
    <p wire:loading.delay>Updating...</p>
</div>
```

And setting up a route for `/` to load a page with this component on it.

Running from `artisan serve` and visiting `localhost:8000` the first click takes around 20ms, the next takes 330ms, the third takes 20ms but the _Updating..._ message stays.

Strangely visiting `127.0.0.1:8000` every click takes about 20ms and everything works as expected, you never see the _Updating..._ message.

**Comment by SeriousKen** [comment]
Done some more looking into this. Firstly, the long request (>200ms) is when the initial connect is about 300ms, the actual response after connect is quite quick - just wondering if the logic is wrong somewhere for when there is a delay in the TCP connection stage.

I've noticed when debugging the livewire.js that the problem seems to occur in the `toggleBooleanStateDirective` function. This function is called twice when processing the click in the above component, the first time when it works correctly `isTruthy` is `true`, then later it becomes `false`. When it fails, the values `isTruth` of the first and second calls are flipped - the first `false`, the second `true` - this causes the message to remain displayed. I'm just trying at the moment to figure out where the erroneous values come from.

**Comment by SeriousKen** [comment]
There seems to be a situation where the callbacks set up by `applyDelay` get triggered in the wrong order. The code here https://github.com/livewire/livewire/blob/ab75c265a9dcbfa92af81be315aed714ec22676a/js/directives/wire-loading.js#L46-L61 the callback inside the timeout on line 49 gets called _after_ the callback on line 56, causing the elements with `loading.delay` to be visible when they shouldn't be.

**Comment by stebers** [comment]
I would also like to see a bug fix for loading delay.

For the meantime, I'd like to share my workaround:

```blade
<script>
  function throttle(func, delay) {
    let lastCall = 0

    return function (...args) {
      const now = new Date().getTime();
      if (now - lastCall >= delay) {
        func(...args)
        lastCall = now
      }
    }
  }

  document.addEventListener('livewire:init', () => {
    Livewire.hook('request', () => {
      const loadingElements = document.querySelectorAll('[custom\\:loading]')

      loadingElements.forEach(e => {
        e.classList.add(e.getAttribute('custom:loading.prevent'))
        e.classList.add(e.getAttribute('custom:loading'))
      })

      setTimeout(() => {
        loadingElements.forEach(e => {
          e.classList.remove(e.getAttribute('custom:loading.prevent'))
        })
      }, 200)
    })

    Livewire.hook('morph.added', throttle(() => {
      const loadingElements = document.querySelectorAll('[custom\\:loading]')
      
      loadingElements.forEach(e => {
        e.classList.remove(e.getAttribute('custom:loading'))
      })
    }, 50))
  })
</script>

<div custom:loading="opacity-50" custom:loading.prevent="!opacity-100">
    ...
</div>
```

**Comment by joalea** [comment]
Was looking all over for this post as I have the exact same problem after upgrading from LW2->LW3.
As a workaround I have removed "delay" form all my loading components although its not as pretty.

Definitely a bug, thanks for addressing it!

**Reply by austenc** [reply]
Just encountered this too -- there's some sort of bug with `.delay` where the `style="display: inline-block;"` doesn't get removed in some cases. It's happening in our case when a `wire:model` input has something entered in it, and then the input is cleared (but the loading state remains).

Haven't figured out the underlying cause just yet, will edit this response if I do!

**Comment by Briareos17** [comment]
Has same on release version v3.0.10, set `wire:loading.delay.longest` for test, component loads and then loader appears and do not disappear

**Comment by austenc** [comment]
I've added a failing test case for this in https://github.com/livewire/livewire/pull/7118

It happens when the first request takes longer than the delay, but the second request **doesn't**. Thanks to @MatinMN for the reproduction code that led to discovering this. ~~Working on a fix now!~~

Edit: PR now includes a fix for this case.

**Reply by rafalpietrzakio** [reply]
Hi! I've just found the thread. I'm running livewire 3.4.5 (the newest one) and the  issue still appears unfortunately. Appears on file uploads (only when wire:loading is delayed in any way). This seem to be the same that I've posted here (together with repo and reproduction steps): https://github.com/livewire/livewire/discussions/7891

I'm reaching you here as you probably addressed the similar issue in some way; this might be the same bug but in different place now

**Comment by rafalpietrzakio** [comment]
Any chance to get this fixed? It happens for all loaders without target
